pandas/io/excel/_odfreader.py

```diff
@@ -0,0 +1,192 @@
+from typing import Dict, List
+
+from pandas.compat._optional import import_optional_dependency
+
+import pandas as pd
+from pandas._typing import FilePathOrBuffer, Scalar
+
+from pandas.io.excel._base import _BaseExcelReader
+
+
+class _ODFReader(_BaseExcelReader):
+    """Read tables out of OpenDocument formatted files
+
+    Parameters
+    ----------
+    filepath_or_buffer: string, path to be parsed or
+        an open readable stream.
+    """
+    def __init__(self, filepath_or_buffer: FilePathOrBuffer):
+        import_optional_dependency("odf")
+        super().__init__(filepath_or_buffer)
+
+    @property
+    def _workbook_class(self):
+        from odf.opendocument import OpenDocument
+        return OpenDocument
+
+    def load_workbook(self, filepath_or_buffer: FilePathOrBuffer):
+        from odf.opendocument import load
+        return load(filepath_or_buffer)
+
+    @property
+    def sheet_names(self) -> List[str]:
+        """Return a list of sheet names present in the document"""
+        from odf.table import Table
+
+        tables = self.book.getElementsByType(Table)
+        return [t.getAttribute("name") for t in tables]
+
+    def get_sheet_by_index(self, index: int):
+        from odf.table import Table
+        tables = self.book.getElementsByType(Table)
+        return tables[index]
+
+    def get_sheet_by_name(self, name: str):
+        from odf.table import Table
+
+        tables = self.book.getElementsByType(Table)
+
+        for table in tables:
+            if table.getAttribute("name") == name:
+                return table
+
+        raise ValueError("sheet {name} not found".format(name))
+
+    def get_sheet_data(self, sheet, convert_float: bool) -> List[List[Scalar]]:
+        """Parse an ODF Table into a list of lists
+        """
+        from odf.table import TableCell, TableRow
+
+        sheet_rows = sheet.getElementsByType(TableRow)
+        table = []  # type: List[List[Scalar]]
+        empty_rows = 0
+        max_row_len = 0
+        row_spans = {}  # type: Dict[int, int]
+
+        for i, sheet_row in enumerate(sheet_rows):
+            sheet_cells = sheet_row.getElementsByType(TableCell)
+            empty_cells = 0
+            table_row = []  # type: List[Scalar]
+
```

Comment:
may want to break this up with some helper methods (e.g. leave the nested loops here ) and call into a function for a cell

Reply:
The tricky thing with moving the cell-level code out is going to be the `row_span` (which signals if a particular cell is merged vertically). This gets passed to a dict whose scope is outside of the loops so that the next row iteration can know if it needs to account for a vertical merge while de-sparsifying

So if the cell-level parsing gets moved to a separate function I think we'd want to return the row_span from that or pass all of the mutable structures into that function. Not sure if that helps readability

^ not objecting just a point to consider

Reply:
Ignore this comment - found a better way

Reply:
Refactor up so hope this is less of an issue. The inner loops both modify containers declared outside of their respective loops so I *think* moving might make things less readable but lmk